### PR TITLE
feat: File row splits

### PR DESCRIPTION
**Pull request by AdamGS**
Introduces a new `VortexReadHandle` that can be used cloned and used to either generate a full stream over the underlying file OR read a specific row range.
The handle also implements `Clone` so it can be re-used over the same file.

DataFusion support will be in a followup PR.

### CODE REVIEW

File: vortex-file/src/read/buffered.rs [hunk 1]
```diff
@@ -26,11 +26,11 @@ pub(crate) trait ReadMasked {
 
 /// Read an array with a [`RowMask`].
 pub(crate) struct ReadArray {
-    layout: Box<dyn LayoutReader>,
+    layout: Arc<dyn LayoutReader>,
```
Comment (gatesn) [inline vortex-file/src/read/buffered.rs:29]: So LayoutReaders are no longer stateful?

File: vortex-file/src/read/handle.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        self.row_count
+    }
+
+    /// Returns a set of row splits in the file, that can be used to inform on how to split it horizontally.
+    pub fn splits(&self) -> &[(usize, usize)] {
+        &self.splits
+    }
+
+    /// Create a stream over all data from the handle
+    pub fn into_stream(self) -> VortexResult<VortexReadArrayStream<R>> {
```
Comment (gatesn) [inline vortex-file/src/read/handle.rs]: Why would this fail? No result?

File: vortex-file/src/read/handle.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        let splits_vec = Vec::from_iter(self.splits().iter().copied());
+        let split_accumulator = SplitsAccumulator::new(splits_vec.into_iter(), self.row_mask);
+
+        let splits_stream = stream::iter(split_accumulator);
+
+        // Set up a stream of RowMask that result from applying a filter expression over the file.
+        let mask_iterator = if let Some(fr) = &self.filter_reader {
+            Box::new(BufferedLayoutReader::new(
```
Comment (gatesn) [inline vortex-file/src/read/handle.rs:88]: Ideally we pull the splits_stream out of this thing and just `stream::iter(self.splits().iter()).map(|s| self.read_range(s).boxed()).buffered(64)` or whatever?

Or is that not possible?

**Review by gatesn — COMMENTED** [review]
This seems like a step in the right direction. I still think we should pull apart BufferedReader thingy.

And are you sure all the state machines have gone from layout readers?

File: vortex-file/src/read/buffered.rs (hunk 1 above)
Comment (AdamGS) [reply vortex-file/src/read/buffered.rs:29]: Seems like I overlooked some layers of of state/locks in the readers, that will take some thinking and probably additional refactoring.

File: vortex-file/src/read/handle.rs (hunk 2 above)
Comment (AdamGS) [reply vortex-file/src/read/handle.rs]: just a leftover, I'll fix that

File: vortex-file/src/read/handle.rs (hunk 3 above)
Comment (AdamGS) [reply vortex-file/src/read/handle.rs:88]: Honestly worried about using `buffered` everywhere

File: vortex-file/src/read/buffered.rs (hunk 1 above)
Comment (AdamGS) [reply vortex-file/src/read/buffered.rs:29]: ok after a more careful reading and a small test, as far as I can tell while the `LayoutReader` implementations are stateful (with interior mutability), they can read independently from each other as the `LayoutBufferedReader` is the part that actually controls the overall flow and each stream initializes a new one.

Comment (a10y) [reply vortex-file/src/read/buffered.rs:29]: There can be multiple concurrent reads being tracked in the layout reader but are they reusable?

I thought something encoded assumptions that splits were accessed monotonically but maybe that was in the BufferedReader

Comment (AdamGS) [reply vortex-file/src/read/buffered.rs:29]: I thought so too, but that does seem to be in the `BufferedLayoutReader`. The `LayoutReader` implementation read through `poll_read` which takes a `RowMask` to read try and read a specific range on each call, and the `BufferedReader` tracks that.

Comment (robert3005) [reply vortex-file/src/read/buffered.rs:29]: You should be able to access the layoutreaders out of order so we can read multiple at the same time irrespective of their children.

File: vortex-file/src/read/handle.rs (hunk 3 above)
Comment (robert3005) [reply vortex-file/src/read/handle.rs:88]: Buffered is a bit too naive. You want more control over coalescing

File: vortex-file/src/read/handle.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+#[derive(Clone)]
+pub struct VortexReadHandle<R> {
+    input: R,
+    dtype: Arc<LazyDType>,
+    row_count: u64,
+    splits: Vec<(usize, usize)>,
+    layout_reader: Arc<dyn LayoutReader>,
+    filter_reader: Option<Arc<dyn LayoutReader>>,
+    messages_cache: Arc<RwLock<LayoutMessageCache>>,
+    row_mask: Option<RowMask>,
```
Comment (a10y) [inline vortex-file/src/read/handle.rs:24]: is it necessary to provide a handle-level row-mask? This sort of feels like a parameter you pass when you build a scan from the handle

**Review by a10y — APPROVED** [review]

File: vortex-file/src/read/handle.rs (hunk 4 above)
Comment (AdamGS) [reply vortex-file/src/read/handle.rs:24]: The flow I imagine is something like:
1. Read some external index to create initial `RowMask`
2. Pass that row to the the builder to create a handle.
3. Reuse the initial mask to create concurrent reads on subranges of the file

**Review by gatesn — APPROVED** [review]

**Pull request merged**